### PR TITLE
using default nwb metadata as base

### PR DESCRIPTION
**Pull request by Saksham20**
THis branch has some changes to the `NwbImagingExtractor.write_nwb()` and `NwbSegmentationExtractor.write_nwb()` and linked methods to reduce redundancies. Lets discuss these on this PR.
1. `NwbSegmentationExtractor` had a method to create a default nwb metadata dictionary before using the specific `SegmentationExtractor` to update it. I've added this method outside the class so that `NwbImagingExtractor` can also leverage this and update those default fields related to Imaging (`ImagingPlane`, `TwoPhotonSeries`, `Device`) 
2. I've also added `NwbImagingExtractor.get_nwb_metadata()` just like the previous `NwbSegmentationExtractor.get_nwb_metadata()` that takes that default nwb metadata dictionary and populates that with the some details from the specific imaging extractor object.

### CODE REVIEW

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 1]
```diff
@@ -89,6 +89,29 @@ def update_dict(d, u):
     return d
 
 
+def get_default_nwb_metadata():
```
Comment (Saksham20) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:92]: default metadata which will be over-ridden.

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 2]
```diff
@@ -256,67 +276,39 @@ def add_two_photon_series(imaging, nwbfile, metadata, num_chunks=10):
...
-                metadata['Ophys']['ImagingPlane'] = [{'name': 'ImagingPlane',
-                                                      'description': 'no description',
-                                                      'excitation_lambda': np.nan,
-                                                      'indicator': 'unknown',
-                                                      'location': 'unknown'}]
-            imaging_meta = {'optical_channel': optical_channel,
-                            'imaging_rate': rate,
-                            'device': device}
-            metadata['Ophys']['ImagingPlane'][0] = update_dict(metadata['Ophys']['ImagingPlane'][0], imaging_meta)
+            metadata['Ophys']['ImagingPlane'][0]['optical_channel'] = \
```
Comment (Saksham20) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:287]: updating metadata with a default dictionary of metadata (`get_default_nwb_metadata()`) helps in removing some lines of code here

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 3]
```diff
@@ -350,6 +342,40 @@ def add_epochs(imaging, nwbfile):
 
         return nwbfile
 
+    @staticmethod
```
Comment (Saksham20) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:345]: this method just overwrites some of the default fields like channel_names, imaging_rate, and image_size (for TwoPhotonSeries)

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 4]
```diff
@@ -89,6 +89,29 @@ def update_dict(d, u):
     return d
 
 
+def get_default_nwb_metadata():
+    metadata = {'NWBFile': {'session_start_time': datetime.now(),
```
Comment (bendichter) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:93]:
```suggestion
    metadata = {'NWBFile': {'session_start_time': datetime(1970,1,1),
```

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 5]
```diff
@@ -256,67 +276,39 @@ def add_two_photon_series(imaging, nwbfile, metadata, num_chunks=10):
...
-                imaging_plane=imaging_plane,
-                rate=rate,
-                unit='normalized amplitude',
-                comments='Generated from RoiInterface::NwbImagingExtractor',
-                description='no description'
-            )
+            metadata['Ophys']['TwoPhotonSeries'][0] = update_dict(metadata['Ophys']['TwoPhotonSeries'][0],
+                                                                  dict(data=data,
+                                                                       imaging_plane=imaging_plane))
+            ophys_ts = TwoPhotonSeries(**metadata['Ophys']['TwoPhotonSeries'][0])
```
Comment (bendichter) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:311]:
```suggestion
            two_p_series_kwargs = update_dict(
                metadata['Ophys']['TwoPhotonSeries'][0],
                dict(
                    data=data,
                    imaging_plane=imaging_plane)
                )
            ophys_ts = TwoPhotonSeries(**two_p_series_kwargs)
```

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 6]
```diff
@@ -350,6 +342,40 @@ def add_epochs(imaging, nwbfile):
...
+        for i in range(imgextractor.get_num_channels()):
+            ch_name = imgextractor.get_channel_names()[i]
+            if i == 0:
+                metadata['Ophys']['ImagingPlane'][0]['optical_channel'][i]['name'] = ch_name
+            else:
+                metadata['Ophys']['ImagingPlane'][0]['optical_channel'].append(dict(
+                    name=ch_name,
+                    emission_lambda=np.nan,
+                    description=f'{ch_name} description'
+                ))
```
Comment (bendichter) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:364]: I don't understand this logic. Why treat the first element differently?

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 7]
```diff
@@ -350,6 +342,40 @@ def add_epochs(imaging, nwbfile):
...
+
+        # set imaging plane rate:
+        rate = np.float(
+            'NaN') if imgextractor.get_sampling_frequency() is None else imgextractor.get_sampling_frequency()
```
Comment (bendichter) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:368]: This is fine, but the formatting is a bit odd

File: roiextractors/extractors/nwbextractors/nwbextractors.py [hunk 8]
```diff
@@ -350,6 +342,40 @@ def add_epochs(imaging, nwbfile):
...
+        # adding imaging_rate:
+        metadata['Ophys']['ImagingPlane'][0].update(imaging_rate=rate)
+        # TwoPhotonSeries update:
+        metadata['Ophys']['TwoPhotonSeries'][0].update(
+            dimension=imgextractor.get_image_size())
+        # remove what Segmentation extractor will input:
+        _ = metadata['Ophys'].pop('ImageSegmentation')
+        _ = metadata['Ophys'].pop('Fluorescence')
```
Comment (bendichter) [inline roiextractors/extractors/nwbextractors/nwbextractors.py:376]: Seems odd to add these as default and then remove them. If you only want them for Segmentation extractor, then add them there

**Comment by bendichter** [comment]
@Saksham20 looks good. A few minor things, and can you resolve the merge conflict as well?

**Review by bendichter — CHANGES_REQUESTED** [review]
looks good. A few minor things, and can you resolve the merge conflict as well?

File: roiextractors/extractors/nwbextractors/nwbextractors.py (hunk 8 above)
Comment (Saksham20) [reply roiextractors/extractors/nwbextractors/nwbextractors.py:379]: there are fewer lines of code and I feel its more readable to remove unneeded fields than add the whole dict to a field (fewer lines)

**Pull request merged**